src/qibojit/backends/clifford_operations_cpu.py

```diff
@@ -0,0 +1,301 @@
+import numpy as np
+from numba import njit, prange, uint64
+
+
+@njit("b1[:,:](b1[:,:], u8, u8)", parallel=True, cache=True)
+def H(symplectic_matrix, q, nqubits):
+    r = symplectic_matrix[:-1, -1]
+    x = symplectic_matrix[:-1, :nqubits]
+    z = symplectic_matrix[:-1, nqubits:-1]
+    for i in prange(symplectic_matrix.shape[0] - 1):  # pylint: disable=not-an-iterable
+        symplectic_matrix[i, -1] = r[i] ^ (x[i, q] & z[i, q])
+        tmp = symplectic_matrix[i, q]
+        symplectic_matrix[i, q] = symplectic_matrix[i, nqubits + q]
+        symplectic_matrix[i, nqubits + q] = tmp
```

Comment:
Here you're essentially working indexing everything with `i`, but never accessing any other rows (first dimension).

So, why are you doing an explicit iteration on `i`? You could just rely on the vectorized operation.

Reply:
This is my understanding of how to parallelize operations with `numba` [(https://numba.pydata.org/numba-doc/latest/user/parallel.html#explicit-parallel-loops)](https://numba.pydata.org/numba-doc/latest/user/parallel.html#explicit-parallel-loops) . I might be wrong, but I am not sure whether
```
symplectic_matrix[:, -1] = r ^ (x[:, q] & z[:, q])
```
is going to be parallelized automatically.

Reply:
> Some operations inside a user defined function, e.g. adding a scalar value to an array, are known to have parallel semantics. A user program may contain many such operations and while each operation could be parallelized individually, such an approach often has lackluster performance due to poor cache behavior. Instead, with auto-parallelization, Numba attempts to identify such operations in a user program, and fuse adjacent ones together, to form one or more kernels that are automatically run in parallel.

and	

> In this section, we give a list of all the array operations that have parallel semantics and for which we attempt to parallelize.
>
> 1. All numba array operations that are supported by [Case study: Array Expressions](https://numba.pydata.org/numba-doc/latest/developer/rewrites.html#case-study-array-expressions), which include common arithmetic functions between Numpy arrays, and between arrays and scalars, as well as Numpy ufuncs. They are often called element-wise or point-wise array operations:
>
>    - unary operators: + - ~
>    -  binary operators: + - * / /? % | >> ^ << & ** //
>    -  comparison operators: == != < <= > >=

From the page you cited

Reply:
mmmmhhh, you were right, it's way faster without the explicit loop....

Reply:
Edit: actually this morning I was doing some further testing: I changed the CNOT and CZ by removing the loop but for some reason the circuit execution is way slower. The CNOT test I did yesterday was always on the initial state |00...0> and, maybe, that particular case yielded better results for some reason...